### PR TITLE
Standardise solidity version to 0.8.4

### PR DESCRIPTION
**Pull request by dimpar**
Closes https://github.com/keep-network/coverage-pools/issues/98

For Solidity upgrade to 0.8.4 this PR also upgrades OZ to 4.1.0 to align with Solidity compiler. Some of the compilation errors had to be changed as well.

### CODE REVIEW

File: contracts/AssetPool.sol [hunk 1]
```diff
@@ -12,11 +12,11 @@
...
 // SPDX-License-Identifier: MIT
 
-pragma solidity <0.9.0;
+pragma solidity 0.8.4;
 
 import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
-import "@openzeppelin/contracts/token/ERC20/SafeERC20.sol";
-import "@openzeppelin/contracts/math/SafeMath.sol";
+import "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import "@openzeppelin/contracts/utils/math/SafeMath.sol";
```
Comment (pdyraga) [inline contracts/AssetPool.sol:19]: Do we still need it?

Comment (dimpar) [reply contracts/AssetPool.sol]: Nope, [Solidity docs](https://docs.soliditylang.org/en/latest/080-breaking-changes.html#how-to-update-your-code) tells to replace it. Done in c154802
However, one file is still untouched `ERC20WithPermit.sol` which includes custom error messages for `.sub()`. Do we want to refactor that as well along with the tests?

File: contracts/ERC20WithPermit.sol [hunk 2]
```diff
@@ -12,9 +12,9 @@
 
 // SPDX-License-Identifier: MIT
 
-pragma solidity <0.8.0;
+pragma solidity 0.8.4;
 
-import "@openzeppelin/contracts/math/SafeMath.sol";
+import "@openzeppelin/contracts/utils/math/SafeMath.sol";
```
Comment (tomaszslabon) [inline contracts/ERC20WithPermit.sol:17]: I `SafeMath` still needed in `ERC20WithPermit.sol`?
I see that `add()` is still used, but it could probably be replaced with `+`.

File: contracts/AssetPool.sol (hunk 1 above)
Comment (pdyraga) [reply contracts/AssetPool.sol]: https://github.com/keep-network/coverage-pools/pull/107/commits/eb8d0546c3bf58d1b065baee638d7f40d830b414 😜 

We do want to keep the same revert messages. I started looking into it, then it escalated... and I figured I'll just commit and leave the review of this change to you.

**Comment by dimpar** [comment]
@pdyraga all the comments related to operators are addressed here: 79af193

> Can we remove uint256 from this conversion?

I've tried that and all the tests passed, but... according to the docs: 

> `address(uint)` and `uint(address)` : converting both type-category and width. Replace this by `address(uint160(uint))` and `uint(uint160(address))` respectively.

 I'm just not sure if there can be some weird corner case where having just `tbtcDepositToken.exists(uint160(depositAddress))` breaks things.

**Comment by pdyraga** [comment]
> > Can we remove uint256 from this conversion?
> 
> I've tried that and all the tests passed, but... according to the docs:
> 
> > `address(uint)` and `uint(address)` : converting both type-category and width. Replace this by `address(uint160(uint))` and `uint(uint160(address))` respectively.
> 
> I'm just not sure if there can be some weird corner case where having just `tbtcDepositToken.exists(uint160(depositAddress))` breaks things.

I think it is safe, it's just a matter of whether `uint160 -> uint256` conversion happens implicitly or explicitly. We properly updated `uint(address)` to `uint(uint160(address))` per Solidity 8 changelog but the additional `uint256` should have nothing to do about Solidity 8. That's fine, it's just a detail, we can leave it as-is.

**Review by pdyraga — APPROVED** [review]
